File: #33_Search_in_Rotated_Sorted_Array/Solution.java
public class Solution {
    public int search(int[] nums, int target) {
    	if (nums.length == 0) return -1;
    	int lo = 0; int hi = nums.length;
    	return search(nums, target, lo, hi - 1);
    }

    private int search(int[] nums, int target, int lo, int hi) {
    	if (nums.length == 0) return -1;
    	if (lo > hi) return -1;
    	int mid = (lo + hi) / 2;
    	if (nums[mid] == target) return mid;
    	/* on the left */
    	if (nums[mid] < nums[lo]) {
    		if (target > nums[mid] && target <= nums[hi]) return search(nums, target, mid + 1, hi);
    		return search(nums, target, lo, mid - 1);
    	}
    	/* on the right */
    	if (nums[mid] > nums[hi]) {
    		if (target < nums[mid] && target >= nums[lo]) return search(nums, target, lo, mid - 1);
    		return search(nums, target, mid + 1, hi);
    	}
    	if (target < nums[mid]) return search(nums, target, lo, mid - 1);
    	return search(nums, target, mid + 1, hi);
    }

    public static void main(String[] args) {
    	int[] test = {4, 5, 6, 7, 0, 1, 2};
    	Solution sol = new Solution();
    	System.out.println(sol.search(test, 5));
    }
}

File: #33_Search_in_Rotated_Sorted_Array/Solution.py
'''
Problems:
1. return index, so must keep the full array
2. after rotation, at least half of the array will remain sorted
3. constraint of i, j, deal with lo == hi, lo == mid while lo != hi
'''

class Solution:
	# @param {integer[]} nums
	# @param {integer} target
	# @return {integer}
	def search(self, nums, target):
		return self.searchInRange(nums, target, 0, len(nums) - 1)

	def searchInRange(self, nums, target, lo, hi):
		if not nums or lo < 0 or hi >= len(nums) or lo > hi:
			return -1
		mid = lo + (hi - lo) / 2 # avoid overflow
		if nums[mid] == target:
			return mid
		if lo == hi:
			return -1
		if mid == lo:
			return self.searchInRange(nums, target, mid + 1, hi)
		# if mid == hi:
		# 	return self.searchInRange(nums, target, lo, mid - 1)
		if nums[mid] > nums[lo]: # left part is sorted
			if nums[lo] <= target and target <= nums[mid - 1]:
				return self.searchInRange(nums, target, lo, mid - 1)
			return self.searchInRange(nums, target, mid + 1, hi)
		if nums[mid] < nums[hi]: # right part is sorted
			if nums[mid + 1] <= target and target <= nums[hi]:
				return self.searchInRange(nums, target, mid + 1, hi)
			return self.searchInRange(nums, target, lo, mid - 1)

sol = Solution()
t = [3, 1]
print sol.search(t, 1)